Add NormalSubgroups methods for symmetric and alternating permutation groups

mtorpey opened the pull request:
The normal subgroups of a symmetric group are well-known and simply described: the trivial group, the Klein 4-group (if n=4), the alternating group, and the symmetric group itself.  An alternating group's normal subgroups are even simpler: the same but without the symmetric group.  The `NormalSubgroups` attribute in GAP should reflect this knowledge by having special methods for permutation groups that are known to be symmetric or alternating.

At the moment, some laborious work is done which means that calling, for example, `NormalSubgroups(SymmetricGroup(40))`, takes several seconds, while larger examples run for a very long time.  This PR introduces simple methods that return the appropriate subgroups.

Note that this only applies to *permutation* groups, since it is easy to describe subgroups based on their moved points.  Note also that this is **not** a fix for Issue #2683, a bug that probably applies to non-symmetric groups as well.

fingolfin commented on lib/gpprmsya.gi:
```diff
@@ -2447,6 +2447,42 @@ InstallMethod( RadicalGroup, "symmetric", true,
+  local pts, list;
+  pts := MovedPoints(G);
+  list := [Group(())];  # Trivial group
+  if Length(pts) = 4 then  # Klein 4-group
```
This check is only valid if the symmetric group is in its "natural" form, as tested by `IsNaturalSymmetricGroup`. But  `Sym(4)` also has e.g. a permutation representation acting on 24 points (i.e. the regular module):
```
gap> G:=Image(RegularActionHomomorphism(SymmetricGroup(4)));;
gap> IsSymmetricGroup(G);
true
gap> IsNaturalSymmetricGroup(G);
false
gap> NrMovedPoints(G);
24
```

fingolfin commented on lib/gpprmsya.gi:
```diff
@@ -2447,6 +2447,42 @@ InstallMethod( RadicalGroup, "symmetric", true,
+
+InstallMethod(NormalSubgroups,
+"for a perm group isomorphic to the alternating group",
+[IsPermGroup and IsAlternatingGroup],
```
I guess something similar could also be installed for `IsSimpleGroup`... Not saying you should, just thinking out loud. The case with few moved points are mostly trivial to compute anyway.

fingolfin commented on lib/gpprmsya.gi:
```diff
@@ -2447,6 +2447,42 @@ InstallMethod( RadicalGroup, "symmetric", true,
+  local pts, list;
+  pts := MovedPoints(G);
+  list := [Group(())];
+  if Length(pts) = 4 then
```
Again, this test only is sufficient if you restrict to `IsNaturalAlternatingGroup`. Or switch the condition of this method to `IsSimpleGroup`.

fingolfin commented:
BTW, It might be better to instead modify `NormalSubgroupsCalc` to teach it about these cases. Probably @hulpke has something to say about that, too... My personal impression here is that this is something where group recognition would help for the full "proper" solution, but @hulpke will know far better.

mtorpey commented:
@ChrisJefferson, you're right.  Amended.

mtorpey commented:
In response to @fingolfin's comments, I've gone for `IsNaturalSymmetric/AlternatingGroup` as the filter since, as he points out, this algorithm only applies to groups that are the symmetric group on their set of moved points.  I've also made the tests less display-focused, which should allow them to pass regardless of which packages are loaded.

I don't know anything about `NormalSubgroupsCalc`, so I'll let others consider that. :)

markuspf commented:
If a group knows `IsSymmetric`, using `DerivedSubgroup` as the one proper normal subgroup would suffice (indeed to compute `IsSymmetric`, GAP computes the derived subgroup of `G`),

randomised recognition I can only find in `recog`.

Adding an additional case to `NormalSubgroupCalc` seems to be a rather complicated fix to this fairly simple problem.

fingolfin commented:
@markuspf I am not sure I understand what you are trying to say with the first paragraph of your last comment. Note that Sym(4) has another normal subgroup besides it derived subgroup.

markuspf commented:
@fingolfin, I was ignoring "trivial" cases.

Let's wait for @hulpke to comment.

fingolfin commented:
Observation: one could simply install `DerivedSeries` as a method in all these cases (does not require "natural" either)

fingolfin commented:
(Well, almost: the trivial group needs to be added in the nonsolvable cases)

markuspf commented:
Ok, now I see why one would want to integrate this into the `NormalSubgroupsCalc`: It'd work for *any* permutation group, even if it doesn't know its natural symmetric (or with improvements isomorphic to a symmetric group).

I think generalising the proposed code to at least work for any group that knows its symmetric is a fairly good thing to do (as @hulpke says this code wouldn't interfere with more sophisticated code that works for examples that don't even know that they're symmetric)

Furthermore, the test for `IsSymmetric` seems fairly cheap, essentially computing the derived subgroup and excluding some pesky counterexamples in the case of `n=6`, but now I am probably veering towards what the code in `NormalSubgroupsCalc` would do anyway.

Incidentally, @hulpke, is here a reference/publication that describes how `NormalSubgroupsCalc` works? That function looks quite daunting at first glance.

mtorpey commented:
I've updated this to use `DerivedSeries` and `DerivedSubgroup`, and therefore to work even for non-natural symmetric and alternating groups.  For now I've left everything in special methods, since I don't know a lot about `NormalSubgroupsCalc`.

In order to have this method selected over the method for `IsPermGroup` (which redirects to `NormalSubgroupsCalc`), I've added `RankFilter(IsPermGroup)` to the methods' priority.  Does this seem reasonable?

@hulpke: Perhaps I went a little overboard writing tests.  I'll rein it in a bit in future! :)

fingolfin commented on tst/testinstall/opers/NormalSubgroups.tst:
```diff
@@ -0,0 +1,113 @@
+
+# Non-natural symmetric groups
+gap> S4 := Group((1,2,3,4), (1,2));;
+gap> hom := ActionHomomorphism(g, Arrangements([1..4], 4), OnTuples);;
```
change `g` to `S4` here (else the test fails when run on its own).

mtorpey replied:
Thank you!  It's always the little things...

fingolfin commented:
I actually think that the tests added here are very good, and wish we had similarly thorough tests for many more operands. In particular, they cover the cases where the set we act on is *not* `[1..n]` (which is often forgotten), deals with edge cases (here: small n) etc.. However, the whole test runs in 250ms on my laptop, which is indeed a bit long (I mean, it's not much, but if you do the same thing for 1000 functions, that turns into 4 minutes). Simple solution: drop the two tests for degree 100 (or replace them by, say, degree 15), then it goes down to 20 milliseconds. Sure, those cases were what motivated you originally, but if we wanted to check for performance regressions, we really need to create a dedicated test setup for that first (see also issue #260).

Anyway: tests are failing because the changes here affect the example in the manual, in `lib/grp.gd:1870`.

hulpke commented:
@markuspf 
The algorithm in `NormalSubgroupsCalc` is described in  my 1998 ISSAC paper (
http://www.math.colostate.edu/~hulpke/paper/normalsub.pdf )
